dev-packages/eslint-plugin/rules/runtime-import-check.js

```diff
@@ -0,0 +1,105 @@
+    electronMain: '/electron-main/',
+};
+
+/**
+ * Mapping of folders to the list of folders it should not import from.
+ * @type {[string, string[]][]}
+ */
+const restrictedMapping = [
+    // We start by declaring the allowed imports, we'll negate those later.
+    [folders.common, []],
+    [folders.browser, [folders.common]],
+    [folders.node, [folders.common]],
+    [folders.electronCommon, [folders.common]],
+    [folders.electronBrowser, [folders.electronCommon, folders.browser, folders.common]],
+    [folders.electronNode, [folders.electronCommon, folders.node, folders.common]],
+    [folders.electronMain, [folders.electronCommon, folders.node, folders.common]]
+    // Next we convert the mapping from "allowed" to a list of "restricted" folders.
+].map(([folder, allowed]) => /** @type {[string, string[]]} */(
+    // We want to restrict everything that's either not `folder` nor in the list of allowed folders.
+    [folder, Object.values(folders).filter(f => f !== folder && !allowed.includes(f))]
+));
+
+/** @type {import('eslint').Rule.RuleModule} */
+module.exports = {
+    meta: {
+        type: 'problem',
+        fixable: 'code',
+        docs: {
+            description: 'prevent imports from folders meant for incompatible runtimes.',
+            url: 'https://github.com/eclipse-theia/theia/wiki/Code-Organization'
+        },
+    },
+    create(context) {
+        let relativeFilePath = path.relative(context.getCwd(), context.getFilename());
+        // Normalize the path so we only deal with forward slashes.
+        if (process.platform === 'win32') {
+            relativeFilePath = relativeFilePath.replace(/\\/g, '/');
+        }
+        // Search for a folder following our naming conventions, keep the left-most match.
+        // e.g. `src/electron-node/browser/node/...` should match `electron-node`
+        let lowestIndex = Infinity;
+        /** @type {string[] | undefined} */
+        let restrictedFolders;
+        /** @type {string | undefined} */
+        let matchedFolder;
+        for (const [folder, restricted] of restrictedMapping) {
+            const index = relativeFilePath.indexOf(folder);
+            if (index !== -1 && index < lowestIndex) {
+                restrictedFolders = restricted;
+                matchedFolder = folder;
+                lowestIndex = index;
+            }
+        }
+        // File doesn't follow our naming convention so we'll bail now.
+        if (matchedFolder === undefined) {
+            return {};
+        }
+        return {
+            ImportDeclaration(node) {
+                checkModuleImport(node.source);
+            },
+            TSExternalModuleReference(node) {
+                checkModuleImport(node.expression);
+            },
+        }
+        function checkModuleImport(node) {
+            const module = /** @type {string} */(node.value);
+            if (restrictedFolders.some(restricted => module.includes(restricted))) {
```

Comment:
Would it be worth it to use a `Set` rather than an array to avoid iterating over all restricted folders for every import statement?

Reply:
We can't since we must match one of the paths (e.g. `/browser/`) to see if it is part of the path (the `module` const).

We are doing a substring search, so can't use a set here.